apps/prairielearn/src/lib/server-jobs.ts

```diff
@@ -0,0 +1,195 @@
+  }
+
+  async exec(file: string, args: string[] = [], options: ServerJobExecOptions = {}): Promise<void> {
+    const msg = `$ ${file} ${args.join(' ')}\n`;
+    this.addToOutput(chalk.greenBright(msg));
+    const proc2 = execa(file, args, {
+      ...options,
+      all: true,
+    });
+
+    proc2.all?.setEncoding('utf-8');
+    proc2.all?.on('data', (data) => {
+      this.addToOutput(data);
+    });
+
+    try {
+      await proc2;
+    } finally {
+      // Ensure there is an empty line after all command output.
+      if (!this.output.endsWith('\n\n')) {
+        this.addToOutput('\n');
+      }
+    }
+  }
+
+  /**
+   * Runs the job sequence and returns a Promise that resolves when the job
+   * sequence has completed. The returned promise will not reject if the job
+   * sequence fails.
+   */
+  async execute(fn: ExecutionFunction): Promise<void> {
+    this.checkAndMarkStarted();
+    await this.executeInternal(fn);
+  }
+
+  /**
+   * Identical to {@link execute}, except it doesn't return a Promise. This is
+   * just used as a hint to the caller that they don't need to wait for the job
+   * to finish. Useful for tools like TypeScript and ESLint that ensure that
+   * promises are awaited.
+   */
+  executeInBackground(fn: ExecutionFunction): void {
+    this.checkAndMarkStarted();
+    this.executeInternal(fn);
+  }
+
+  private checkAndMarkStarted() {
+    if (this.started) {
+      throw new Error('ServerJob already started');
+    }
+    this.started = true;
+  }
+
+  private async executeInternal(fn: ExecutionFunction): Promise<void> {
+    try {
+      await fn(this);
+      await this.finish();
+    } catch (err) {
+      try {
+        await this.finish(err);
+      } catch (err) {
+        logger.error(`Error failing job ${this.jobId}`, err);
+        Sentry.captureException(err);
+      }
+    }
+  }
+
+  private addToOutput(msg: string) {
+    this.output += msg;
+    const ansiUp = new AnsiUp();
+    const ansifiedOutput = ansiUp.ansi_to_html(this.output);
+    socketServer.io
+      .to('job-' + this.jobId)
+      .emit('change:output', { job_id: this.jobId, output: ansifiedOutput });
+  }
+
+  private async finish(err: any = undefined) {
+    // Guard against handling job finish more than once.
+    if (this.finished) return;
+    this.finished = true;
+
+    if (err) {
+      // If the error has a stack, it will already include the stringified error.
+      // Otherwise, just use the stringified error.
+      if (err.stack) {
+        this.error(err.stack);
+      } else {
+        this.error(err.toString());
+      }
+
+      if (err.data) {
+        this.verbose('\n' + JSON.stringify(err.data, null, 2));
+      }
+    }
+
+    delete serverJobs.liveJobs[this.jobId];
+
+    await queryAsync(sql.update_job_on_finish, {
+      job_sequence_id: this.jobSequenceId,
+      job_id: this.jobId,
+      output: this.output,
+      status: err ? 'Error' : 'Success',
+      error_message: err?.toString() ?? null,
```

Comment:
@mwest1066 taking a second look at this, do we actually want `error_message` to be a part of non-legacy server jobs? This just repeats information that's already visible in the logs, so I'm leaning towards removing it.

Reply:
Yeah, I agree it's not needed

Reply:
Removed in 02cb07d 🚀